### PR TITLE
livecheck: allow nil return from strategy blocks

### PR DESCRIPTION
**Pull request by samford**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR intends to simplify early returns in a `livecheck` block's `strategy` block. Acceptable return types differ between strategies and it's currently necessary to keep these in mind when using `next` to return early from a `strategy` block.

For example, `PageMatch` (and the strategies that use it internally) will accept a string or array of strings from a `strategy` block, so it's necessary to use something like `next ""` or `next []` when doing an early return.

A recent homebrew/cask PR got me thinking about this because it used a `HeaderMatch` `strategy` block, which can only return a string but will also accept `nil`. Handling a `nil` value means that we can use `next` (with no argument) to return early from a `strategy` block and have the strategy take care of what should happen with a `nil` value.

This behavior seems like an improvement over requiring everyone to remember acceptable return types for a given strategy when we simply want to do an early return. As such, this PR modifies all the strategies to allow for a `nil` return from a `strategy` block, so they can all use `next` for an early return without being required to supply a default value.

Existing enforcement of return types will continue to apply to non-`nil` return values. This PR tries to better align related logic across strategies, where possible. `HeaderMatch` was a notable exception and I've tried to bring that more in line with the others.

---

Past that, this also makes some related changes:

* I've modified the `Git` and `PageMatch` strategies to call `#compact` and `#uniq` on array values returned from a `strategy` block, so we don't have to duplicate this logic in every related `strategy` block. There are a handful of formulae/casks that call `#compact` in a `strategy` block before returning an array and doing this in the strategy itself will allow us to remove those calls and not have to worry about it in the future.
* I've refactored the default `HeaderMatch` logic that checks the `Content-Disposition` and `Location` headers for version information. The existing code was duplicated for each of these headers, so I restructured it to avoid the duplication and to make it easier to add additional headers to check in the future by simply updating an array. This isn't strictly related to this PR and I only added it here because I was already modifying the related code. If necessary, I can split this into a follow-up PR.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense to me. One question: what's the advantage over e.g. always doing `next ""` in those blocks?

**Comment by samford** [comment]
> One question: what's the advantage over e.g. always doing `next ""` in those blocks?

In my view, allowing a `nil` return and standardizing on it slightly reduces the mental load of working on a `strategy` block. String is the common return type from a `strategy` block for the existing strategies but someone who's new to livecheck may not know that and could end up using `next` instead of `next ""` or even some other technically acceptable value (e.g., `next []` for a `PageMatch` or `Git` `strategy` block).

`brew livecheck` would raise an error if it encountered an unacceptable return value from a `strategy` block (e.g., ```Return value of `strategy :page_match` block must be a string or array of strings.```) but I feel like it's natural to want to use `next if ...`. Making that possible for all strategies and standardizing on it allows maintainers/contributors to not have to think about a return type when they just want code that says, "Don't proceed if ...".

Past that, there are some `strategy` blocks where we return early with a meaningful value (e.g., an array of version strings). For example, in `bash` we find major/minor versions on one page and have to check a separate page for patch versions (this is one of a handful of outliers that requires this type of setup). If the second page doesn't provide any patch versions, then we return the major/minor versions (instead of proceeding to the code that works with the patch versions). In this instance, standardizing on using `next if ...` (with no value) will make early returns with a value stand out more.

Long story short, the differences are minor but smoothing rough edges may add up to a better experience over time.

**Comment by samford** [comment]
Looking at the Codecov annotations, I'll aim to add more tests that exercise related code a bit later today.

**Comment by MikeMcQuaid** [comment]
> Long story short, the differences are minor but smoothing rough edges may add up to a better experience over time.

Fine with me, thanks for explaining 👍🏻

**Review by ESA-Security-Group — APPROVED** [review]
#11610

**Comment by samford** [comment]
The latest push adds additional tests where currently possible. I'll have to refactor some of the related `#find_versions` code into separate methods to be able to add more tests but I'll create a follow-up PR for that (I already have part of it finished).

I did final comparative runs with/without these changes and everything looked fine, so this should be good to go at this point.
